Converts ai tracking towards using signals

MNarath1 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

This PR rewrites/refractors ai tracking code towards using signals instead going away from the god awfull spawn(0) while loop.
Most behavior stays the same except that the tracking proc doesn't sleep anymore that was most likely an attempt of the original coder of reducing the actual impact of their while loop but for this its obviously no longer needed so now the ai will no longer lag behind their tracking target.
When we start tracking (and if we can track said target) 2 Signals get registered first in case the target gets somehow qdeleted (this will simply call the stop tracking proc) and obviously a signal for when the target moves that will call the ai_actual_track proc.
Loosing a tracking target now gets handled by a timer that will call the ai_stop_tracking proc if it runs out however the timer can be stopped if the target moves into an area again where it can be tracked by the ai. The timer is set for 10 seconds that should line up with the previous behavior as before it would just be 10 loop iliterations and the loop would sleep for 10 deciseconds every illiteration.
Also moves move_camera_by_click proc to eye object instead of atom.

## Why It's Good For The Game

spawn bad especially if it creates a dedicated while loop

</details>

## Changelog
:cl:
code: converts ai tracking code towards using signals
tweak: the change camera network verb will not immidiatly stop ai tracking(until you select a network)
refactor: moves move_camera_by_click proc to the eyeobject
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

MNarath1 committed:
converts ai tracking towards using signals

this rewrites ai tracking code from its former spawn(0) while loop to using signals instead behaviors stays mostly consistant except that its a bit faster because there is no sleep proc in there anymore

MNarath1 committed:
small change

MNarath1 committed:
whops almost missed that one

MNarath1 committed:
moves move_camera_by_click to eye object

MNarath1 committed:
small change

MNarath1 committed:
fix and implemention for resolving an edge case

MNarath1 committed:
comment change

MNarath1 committed:
changing proc name and small fix

MNarath1 commented:
For the eye object check i am not completly sure if its still needed but according to codebus there might be some exceptions still that would allow for a missing eye object

MNarath1 committed:
removes eyeobj check

i tried the cases where this is supposed to happen but it can't litterally unless some adminbus takes place i don't see any way this happens

MNarath1 commented:
I tried again to see if its actually possible to end up with an eyeobj null but i can't see any way this ever can happen neither can it be deleted nor does it get recreated when you change zlevel etc etc.

MNarath1 committed:
oops

MNarath1 committed:
forgot a space

MNarath1 committed:
puts the return one step higher and inverts the if check

this way i only need a single return instead of two so i guess its better

PowerfulBacon commented on code/game/machinery/camera/tracking.dm:
```diff
@@ -68,57 +68,45 @@
+	eyeobj.setLoc(get_turf(target)) //on the first call of this we obviously need to jump to the target ourselfs else we would go there only after they moved once
+	to_chat(src, "<span class='notice'>Now tracking [target.get_visible_name()] on camera.</span>")
+
+/mob/living/silicon/ai/proc/ai_stop_tracking(var/reacquire_failed = FALSE) //stops ai tracking
```
Since this is being used as a signal handler coming in from RegisterSignal(target, COMSIG_PARENT_QDELETING, .proc/ai_stop_tracking)
the parameter 'reacquire_failed' will actually be datum/source when called from the mob being deleted. This will probably still work, but it should use a wrapper with the correct parameters for COMSIG_PARENT_QDELETING to prevent any weirdness and to make the code cleaner

MNarath1 replied:
True i totally forgot about that part of signal behavior not hard to fix trough

MNarath1 replied:
yeah i only really need to fix the parameter offset since the only time reaquire_failed can and should be true is if it gets called by the CALLBACK timer

MNarath1 replied:
Turns out the COMSIG_PARENT_QDELETING can also send `SEND_SIGNAL(D, COMSIG_PARENT_QDELETING, force) // Let the (remaining) components know about the result of Destroy` the force argument so it needs to that also needs to be taken into account.

MNarath1 replied:
I could also wrap the signal proc but that would still require the same offsets here

PowerfulBacon commented on code/game/machinery/camera/tracking.dm:
```diff
@@ -68,57 +68,45 @@
+	if(reacquire_timer)
+		if(reacquire_failed) //edge case when someone might jump to another camera while the reacquire timer is running
+			to_chat(src, "<span class='warning'>Unable to reacquire, cancelling track...</span>")
+		else
+			deltimer(reacquire_timer)
```
Shouldn't it always delete the timer if you are going to set the timer to null?

MNarath1 replied:
This is a edge case where the timer is running but the tracking got actually interrupted by something else to stop it from calling the proc afterwards trough callback again.

MNarath1 replied:
its not required to delete the timer in the other instance since it will be automatically deleted in that case anyway still need to null the timer id trough

PowerfulBacon reviewed (commented):
Good code quality

MNarath1 committed:
fixes argument offset caused by signals

because signals will forward the reference of the signal source and the qdelete signal can also forward the force argument this needs to be offset by the correct ammount

MNarath1 committed:
changed my mind lets actually wrap this in another proc
